### DISCUSSION TITLE
[Implemented] Specification for `plotObservedVsSimulated()`

### DISCUSSION THREAD

**Discussion by IndrajeetPatil** [Open-Systems-Pharmacology/OSPSuite-R#809] [Ideas]
Related to #674 

# Signature

```r
# `...` will be parameters to control annotations and aesthetics of the plot
plotObservedVsPredicted(dataCombined, ...)
```

# Implementation plan

```r
data <- toObservedVsPredictedDataFrame(dataCombined)

tlf::plotObsVsPred(data, ...) 
```

# Rough expected output:

![image](https://user-images.githubusercontent.com/11330453/157632539-84557ec4-8758-4e40-aa57-7cf45a70b406.png)

# Decision points:

N.B. The option already rejected by the team is scratched out.

- [x] Pairing non-existing data: ~~threshold~~ versus interpolation
- [x] Nature of the needed function to extract needed dataframe: ~~Member method~~ versus function
- [x] Data format returned by respective function: wide versus tidy

# Tasks

Each task will get its own PR.

- Implement `toObservedVsPredictedDataFrame()` function
- Implement `plotObservedVsSimulated()`

**Comment by IndrajeetPatil** [comment]
As for wide versus tidy, you _need_ the data to be in the wide format _irrespective of_ what package you choose, doesn't matter if it's _tlf_, _ggplot2_, or _base_. Paired data that needs to be plotted as a scatterplot will always have to be in wide format. It will be confusing for the users for it _not_ to be.

```r
# tlf
plotObsVsPred(
  data = df,
  dataMapping = ObsVsPredDataMapping$new(
    x = "yValuesObserved",
    y = "yValuesPredicted",
  )
)

# ggplot2

ggplot(df, aes(yValuesObserved, yValuesPredicted)) + 
  geom_point()

# base

plot(df$yValuesObserved, df$yValuesPredicted)
```

**Comment by IndrajeetPatil** [comment]
@PavelBal Should the labels for this plot contain information about dimensions and units, like in profile plots? 
E.g., `Observed values (Fraction [%])` and `Simulated values (Fraction [%])`.

**Comment by PavelBal** [comment]
> Should the labels for this plot contain information about dimensions and units, like in profile plots?

Yes

And let's rename it to `plotObservedVsSimulated`.

**Comment by IndrajeetPatil** [comment]
In profile plots, datasets that are not part of any group are assigned to their own group.

But, in case of the current function, those datasets should be left out of the plot, correct? 

In the example plot, e.g., there are no datasets included in the plot that don't belong to any group.

**Comment by PavelBal** [comment]
Each plotted "data set" is a combination of simulation results and observed data.

So even in case of multiple observed data are assigned to the same group, but no simulatio result, these data are not to be plotted.

The same goes for simulated results - if multiple simulated results are in one group, but no observed data - do not consider.

Maybe like this:

Name | Type | Group
-- | -- | --
Sim1 | Simulated | GroupA
Sim2 | Simulated | GroupA
  |   |  
Obs1 | Observed | GroupB
Obs2 | Observed | GroupB
  |   |  
Sim3 | Simulated | GroupC
Obs3 | Observed | GroupC
  |   |  
Sim4 | Simulated | GroupD
Obs4 | Observed | GroupD
Obs5 | Observed | GroupD
  |   |  
Sim5 | Simulated | GroupE
Sim6 | Simulated | GroupE
Obs7 | Observed | GroupE
  |   |  
Sim7 | Simulated | GroupF
Sim8 | Simulated | GroupF
Obs8 | Observed | GroupF
Obs9 | Observed | GroupF
  |   |  
Sim9 | Simulated |  
  |   |  
Obs10 | Observed |  


Will result in following pairs:

- Sim3 - Obs3
- Sim4 - Obs4
- Sim4 - Obs5
- Sim5 - Obs7
- Sim6 - Obs7
- Sim7 - Obs8
- Sim8 - Obs8
- Sim7 - Obs9
- Sim8 - Obs9

**Comment by msevestre** [comment]
What about a warning when the dataset are being discarded so that the user knows, that something is not quite right?

**Comment by PavelBal** [comment]
I personally do not see a necessity for a warning, but if you think so..

**Reply by msevestre** [reply]
Well the user is doing something that in the context of the function, does not make sense. While I don't think error is necessary, warning can help learning the API better.

**Comment by IndrajeetPatil** [comment]
How about the following?

The `.removeUnpairableDatasets()` will be internally used by `plotObservedVsSimulated()`. 
``` r
library(dplyr)

df <- dplyr::tribble(
  ~name, ~dataType, ~group,
  "Sim1", "Simulated", "GroupA",
  "Sim2", "Simulated", "GroupA",
  "Obs1", "Observed", "GroupB",
  "Obs2", "Observed", "GroupB",
  "Sim3", "Simulated", "GroupC",
  "Obs3", "Observed", "GroupC",
  "Sim4", "Simulated", "GroupD",
  "Obs4", "Observed", "GroupD",
  "Obs5", "Observed", "GroupD",
  "Sim5", "Simulated", "GroupE",
  "Sim6", "Simulated", "GroupE",
  "Obs7", "Observed", "GroupE",
  "Sim7", "Simulated", "GroupF",
  "Sim8", "Simulated", "GroupF",
  "Obs8", "Observed", "GroupF",
  "Obs9", "Observed", "GroupF",
  "Sim9", "Simulated", NA,
  "Obs10", "Observed", NA
)

.removeUnpairableDatasets <- function(data) {
  # How many rows were originally present
  originalDatasets <- unique(data$name)
  
  # Remove datasets that don't belong to any group.
  data <- dplyr::filter(data, !is.na(group))
  
  # Remove groups (and the datasets therein) with only one type (either only
  # observed or only simulated) of dataset.
  data <- data %>%
    dplyr::group_by(group) %>%
    dplyr::filter(length(unique(dataType)) > 1L) %>%
    dplyr::ungroup()
  
  # How many rows are present after filtering
  finalDatasets <- unique(data$name)
  
  # Warn the user about the filtering if it took place
  if (length(finalDatasets) < length(originalDatasets)) {
    missingDatasets <- originalDatasets[!originalDatasets %in% finalDatasets]
    
    message(
      "Following non-grouped or unpairable datasets have been removed:\n",
      paste0(missingDatasets, collapse = "\n")
    )
  }
  
  return(data)
}

df <- .removeUnpairableDatasets(df)
#> Following non-grouped or unpairable datasets have been removed:
#> Sim1
#> Sim2
#> Obs1
#> Obs2
#> Sim9
#> Obs10

df
#> # A tibble: 12 x 3
#>    name  dataType  group 
#>    <chr> <chr>     <chr> 
#>  1 Sim3  Simulated GroupC
#>  2 Obs3  Observed  GroupC
#>  3 Sim4  Simulated GroupD
#>  4 Obs4  Observed  GroupD
#>  5 Obs5  Observed  GroupD
#>  6 Sim5  Simulated GroupE
#>  7 Sim6  Simulated GroupE
#>  8 Obs7  Observed  GroupE
#>  9 Sim7  Simulated GroupF
#> 10 Sim8  Simulated GroupF
#> 11 Obs8  Observed  GroupF
#> 12 Obs9  Observed  GroupF
```

<sup>Created on 2022-05-09 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

Also, to be consistent with `DataCombined`, I am producing a message instead of a warning (cf. https://github.com/Open-Systems-Pharmacology/OSPSuite-R/blob/e2e1e6501c9b45f1a4299cc4b24974a905ba0708/R/data-combined.R#L544-L555).

**Reply by msevestre** [reply]
Message vs warning? What's the difference?
And why not a warning?

**Reply by IndrajeetPatil** [reply]
> Message vs warning? What's the difference?

cf. https://adv-r.hadley.nz/conditions.html#signalling-conditions

<img width="729" alt="Screenshot 2022-05-09 at 15 34 38" src="https://user-images.githubusercontent.com/11330453/167421605-25b50bcd-8349-46c8-affb-b7125259ed9d.png">

> And why not a warning?

Because nothing has gone wrong in these functions. We are just informing users that we did something on their behalf. They could have removed these datasets before using this function. But, since they didn't, we are doing it for them. So this should be a message.

**Reply by msevestre** [reply]
I think it's not a clear line. Something has gone wrong because the function could not process the input. At any rate, this is fine. But make sure to move, as always, the message to the constant files

**Comment by IndrajeetPatil** [comment]
Is this in the ballpark of what you had in mind? 
(The pattern for `distal` group looks odd, but I am not sure if it's due to interpolation not working as expected or because the dataset itself is messy.)

```r
library(ospsuite)

# load the simulation
sim <- loadTestSimulation("MinimalModel")
simResults <- importResultsFromCSV(
  simulation = sim,
  filePaths = getTestDataFilePath("Stevens_2012_placebo_indiv_results.csv")
)

# import observed data (will return a list of DataSet objects)
dataSet <- loadDataSetsFromExcel(
  xlsFilePath = getTestDataFilePath("CompiledDataSetStevens2012.xlsx"),
  importerConfiguration = loadDataImporterConfiguration(getTestDataFilePath("ImporterConfiguration.xml"))
)

# create a new instance and add datasets
myCombDat <- DataCombined$new()
myCombDat$addSimulationResults(
  simResults,
  quantitiesOrPaths = c(
    "Organism|Lumen|Stomach|Dapagliflozin|Gastric retention",
    "Organism|Lumen|Stomach|Dapagliflozin|Gastric emptying",
    "Organism|Lumen|Stomach|Metformin|Gastric retention"
  )
)

myCombDat$addDataSets(dataSet)

myCombDat$setGroups(
  names = c(
    "Organism|Lumen|Stomach|Dapagliflozin|Gastric emptying",
    "Stevens_2012_placebo.Sita_dist",
    "Organism|Lumen|Stomach|Dapagliflozin|Gastric retention",
    "Stevens_2012_placebo.Sita_proximal",
    "Organism|Lumen|Stomach|Metformin|Gastric retention",
    "Stevens_2012_placebo.Sita_total"
  ),
  groups = c("distal", "distal", "proximal", "proximal", "total", "total")
)

# default
plotObservedVsSimulated(myCombDat)
```

![image](https://user-images.githubusercontent.com/11330453/173840019-921a8a9c-b45d-4347-9bf7-d87d173aef06.png)

**Comment by pchelle** [comment]
@PavelBal 
What should be the expected behaviour for __linear__ scaled obs vs pred plots ?

- 1. Keep a fold distance approach with additional lines looking like this
![image](https://user-images.githubusercontent.com/45343665/177353361-b48a19a5-0aa8-429c-b67d-ec3489571ffa.png)

- 2. Using an absolute error instead of fold distance (with potential change of input argument e.g. `errorRange` ?) as below
![image](https://user-images.githubusercontent.com/45343665/177353673-c80cb33a-c534-4844-86e8-495b0d64fef0.png)

**Comment by PavelBal** [comment]
@pchelle i.

**Comment by IndrajeetPatil** [comment]
@pchelle Any update on this? Will you be making these changes in _tlf_ itself? Thanks.
